### PR TITLE
ToolsPanel: Remove hardcoded classnames

### LINKED ISSUE #35058: ToolsPanel: move away from hardcoded classnames
## What

Requires #34345 first, to make sure that tests don't rely on classnames.

We should refactor the `ToolsPanel` component so that it avoids using hardcoded classnames as much as possible.

Currently, the hardcoded classnames in `TooolsPanel` are:

- `components-tools-panel`. This class could be replaced by using the context system, so that `ToolsPanelHeader` knows if it's being rendered inside a `ToolsPanel`
- `single-column`. This class should be removed because it currently introduces circular dependencies with the [`block-editor`](https://github.com/WordPress/gutenberg/blob/4dc6fe86d1f59bab985eb0cb8be481be582415f6/packages/block-editor/src/hooks/dimensions.js#L101) and [`edit-site`](https://github.com/WordPress/gutenberg/blob/20c711d8e75a6dfb12425eacb14f71766a2be3df/packages/edit-site/src/components/sidebar/dimensions-panel.js#L182) (as also specified in https://github.com/WordPress/gutenberg/pull/34725#pullrequestreview-761168248)
- `components-base-control` as explained in https://github.com/WordPress/gutenberg/pull/34725#discussion_r715297703
- `components-dropdown-menu` and `components-dropdown-menu__toggle` — these classnames are going to be tougher to remove, since they would need first the `DropdownMenu` component to be refactored and use the new context System. After that, we could use the component's computed selector at runtime ([example](https://github.com/WordPress/gutenberg/blob/3fd7cdc97d56786a42ce51f9e8738f0c7cb4e6e8/packages/components/src/flyout/styles.ts#L36))

## Why

We should avoid using hardcoded classnames as much as possible in our components:

- it goes against the [Styling guidelines](https://kentcdodds.com/blog/testing-implementation-details), where classnames are generated at runtime
- hardcoded classnames can be effectively treated as public API, which can later need to involuntary breaking changes later on.

## A/C

- [ ] Remove the `components-tools-panel` classname
- [ ] Remove the `single-column` classname
- [ ] Remove the `components-dropdown-menu` and `components-dropdown-menu__toggle` classnames

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Fixes: https://github.com/WordPress/gutenberg/issues/35058

## Description

This PR removes the use of hardcoded CSS classnames from the `ToolsPanel` styles. The one exception is the `single-column` class as this will be removed an a separate PR adopting the Grid component to handle layout for the `ToolsPanel.

## How has this been tested?

Manually.

#### Testing Instructions

1. Checkout this PR and start up Storybook `npm run storybook:dev`
2. Navigate to the `ToolsPanel` story and confirm it still displays and functions correctly
3. On a local site, open the block editor, add a cover block, and select it
4. Confirm the dimensions panel displays and functions correctly

## Screenshots 

Screen Shot 2021-10-07 at 4 09 32 pm

## Types of changes
Enhancement.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by glendaviesnz** [comment]
This tested well for me in Storybook and in local site cover block. The emotion css stuff is new to me though, so will leave it to someone with more experience in that to sign it off.

File: packages/components/src/tools-panel/styles.ts
Comment (ciampo) [inline packages/components/src/tools-panel/styles.ts]: This specificity hack shouldn't be necessary — I tried removing it and the spacing still looks correct (at least in Storybook).

I then noticed that, in the markup, we're using a `h2` and that's we're adding flexbox styles manually. And so, I played around and leveraged the `HStack` and `Heading` components to reduce the amount of custom CSS needed in a component (similarly to what we're going to do by leveraging `Grid` in a future PR).

Using `Heading` this way can also help us improve the overall consistency of the typography system (see related #35464)

<details>
<summary>This is what I came up with</summary>

```diff
diff --git a/packages/components/src/tools-panel/styles.ts b/packages/components/src/tools-panel/styles.ts
index 5135d2287e..6de4d21a5b 100644
--- a/packages/components/src/tools-panel/styles.ts
+++ b/packages/components/src/tools-panel/styles.ts
@@ -57,18 +57,7 @@ export const ToolsPanelHiddenInnerWrapper = css`
 `;
 
 export const ToolsPanelHeader = css`
-	align-items: center;
-	display: flex;
-	font-size: inherit;
-	font-weight: 500;
 	${ toolsPanelGrid.item.fullWidth }
-	justify-content: space-between;
-	line-height: normal;
-
-	/* Required to meet specificity requirements to ensure zero margin */
-	&& {
-		margin: 0;
-	}
 
 	/* Tweak dropdown menu and toggle button for better alignment */
 	> div {
@@ -86,6 +75,12 @@ export const ToolsPanelHeader = css`
 	}
 `;
 
+export const ToolsPanelHeading = css`
+	font-size: inherit;
+	font-weight: 500;
+	line-height: normal;
+`;
+
 export const ToolsPanelItem = css`
 	${ toolsPanelGrid.item.fullWidth }
 
diff --git a/packages/components/src/tools-panel/tools-panel-header/component.tsx b/packages/components/src/tools-panel/tools-panel-header/component.tsx
index 3204615434..0373f6adb7 100644
--- a/packages/components/src/tools-panel/tools-panel-header/component.tsx
+++ b/packages/components/src/tools-panel/tools-panel-header/component.tsx
@@ -16,6 +16,8 @@ import { __, sprintf } from '@wordpress/i18n';
 import DropdownMenu from '../../dropdown-menu';
 import MenuGroup from '../../menu-group';
 import MenuItem from '../../menu-item';
+import { Heading } from '../../heading';
+import { HStack } from '../../h-stack';
 import { useToolsPanelHeader } from './hook';
 import { contextConnect, WordPressComponentProps } from '../../ui/context';
 import type {
@@ -122,6 +124,7 @@ const ToolsPanelHeader = (
 		menuItems,
 		resetAll,
 		toggleItem,
+		headingClassName,
 		...headerProps
 	} = useToolsPanelHeader( props );
 
@@ -133,8 +136,10 @@ const ToolsPanelHeader = (
 	const optionalItems = Object.entries( menuItems?.optional || {} );
 
 	return (
-		<h2 { ...headerProps } ref={ forwardedRef }>
-			{ labelText }
+		<HStack { ...headerProps } ref={ forwardedRef }>
+			<Heading level={ 2 } className={ headingClassName }>
+				{ labelText }
+			</Heading>
 			{ hasMenuItems && (
 				<DropdownMenu
 					icon={ moreVertical }
@@ -168,7 +173,7 @@ const ToolsPanelHeader = (
 					) }
 				</DropdownMenu>
 			) }
-		</h2>
+		</HStack>
 	);
 };
 
diff --git a/packages/components/src/tools-panel/tools-panel-header/hook.ts b/packages/components/src/tools-panel/tools-panel-header/hook.ts
index f31e8cb637..26a2ebca57 100644
--- a/packages/components/src/tools-panel/tools-panel-header/hook.ts
+++ b/packages/components/src/tools-panel/tools-panel-header/hook.ts
@@ -25,6 +25,11 @@ export function useToolsPanelHeader(
 		return cx( styles.ToolsPanelHeader, className );
 	}, [ className ] );
 
+	const headingClassName = useMemo(
+		() => cx( styles.ToolsPanelHeading ),
+		[]
+	);
+
 	const dropdownMenuClassName = useMemo( () => {
 		return cx( styles.DropdownMenu );
 	}, [] );
@@ -33,9 +38,10 @@ export function useToolsPanelHeader(
 
 	return {
 		...otherProps,
-		dropdownMenuClassName,
 		hasMenuItems,
 		menuItems,
 		className: classes,
+		dropdownMenuClassName,
+		headingClassName,
 	};
 }
```

File: packages/components/src/tools-panel/styles.ts
Comment (ciampo) [inline packages/components/src/tools-panel/styles.ts]: I'm tempted to leave these two dropdown-related hardcoded classNames in, for the time being. The correct way to deal with them would be to refactor the `DropdownMenu` component to use Emotion internally (and expose the correct props and/or style selectors).

Maybe we should instead add a comment explaining that this solution is temporary and that we should refactor away from these classes in the near future?

**Review by ciampo — COMMENTED** [review]
Thank you for working on this, @aaronrobertshaw !

I left a couple of comments. I would also love to hear @sarayourfriend 's opinion on this PR and the solutions we've adopting. I'm also cc'ing @mirka for visibility.

File: packages/components/src/tools-panel/styles.ts
Comment (sarayourfriend) [inline packages/components/src/tools-panel/styles.ts]: This might be worse than the classnames in my opinion... because it relies on an even more specific (and completely un-searchable from a text editor perspective) part of the markup. At least with the classname you can pop that into your editor's search functionality and discover what's meant to be targeted. Doing this makes it very difficult to know what the intention is and even easier for something to change in the components that would break this without someone realizing. Again, at least with classnames you might search for usages of it before removing it, but you wouldn't (and shouldn't) think twice about changing the markup of a component because of external styles affecting it.

Comment (ciampo) [reply packages/components/src/tools-panel/styles.ts]: This is a very good point (and explained super clearly!) and in line with what proposed in https://github.com/WordPress/gutenberg/pull/35415#discussion_r725887361. Let's keep the dropdown classnames as they are currently and remove them only after the `Dropdown` menu has been converted to Emotion

File: packages/components/src/tools-panel/styles.ts
Comment (sarayourfriend) [reply packages/components/src/tools-panel/styles.ts]: I agree with this approach... it seems like the most correct one. The best would be if the appropriate components were exported and could be targeted using component interpolation.

Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/styles.ts]: I'm also in agreement it would be best if the `DropdownMenu` components were refactored to use Emotion. 

This current approach was to satisfy the recent request to completely avoid all hardcoded class names. I still think this is worth proceeding with as an interim measure.

I'll add the suggested comments explaining the temporary nature of those styles.

File: packages/components/src/tools-panel/styles.ts
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/styles.ts]: Not a problem. I can revert these changes then. 

It was explained previously, that exposing any hardcoded class names here means we're forever tied to maintaining backwards compatibility for that. Could removing the actual use of the class now but adding a comment with the classname `components-dropdown-menu`, address both concerns?

We then wouldn't have the classname in the styles yet it would still be found when searching for it in an editor as outlined. I don't feel strongly about this, it's just another possible option. What do you think?

File: packages/components/src/tools-panel/styles.ts
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/styles.ts]: Thanks @ciampo for going the extra mile yet again in your reviews, it's appreciated 👍 

The increased specificity was required for use outside of the storybook. Previous tweaks to the heading's styles from within the block editor to overcome its own heading styles were requested to be removed and that the component should ensure the zero margin itself, hence the rule as it stands now.

The use of the `h2` is one of the last remnants of this component [having evolved from a basic clone of the existing panel](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/panel/header.js).

I'll roll your proposed changes into this PR tomorrow when I'll have more time to test it all with real-world use cases such as block supports.

Comment (ciampo) [reply packages/components/src/tools-panel/styles.ts]: > Thanks @ciampo for going the extra mile yet again in your reviews, it's appreciated 👍

It's my pleasure — it's important that we share knowledge as much as possible, especially around the components package and its specific requirements.

> The increased specificity was required for use outside of the storybook. 

Thank you for the context. I suspected there must have been a use case for it 😅 

> I'll roll your proposed changes into this PR tomorrow when I'll have more time to test it all with real-world use cases such as block supports.

No rush as always. I hope these changes will still hold up well in real-world use 🤞

File: packages/components/src/tools-panel/styles.ts
Comment (ciampo) [reply packages/components/src/tools-panel/styles.ts]: That is another way of addressing it, but don't think that it won't make much difference (also because that dropdown hardcoded classname already exists in production).

My favourite approach here is to keep the dropdown classnames as they are currently, and add a comment on top of them as previously discussed

File: packages/components/src/tools-panel/styles.ts
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/styles.ts]: The switch to using `HStack` and `Heading` is a nice step forward. There were two catches with it that I found while testing.

1. The hack to enforce `margin-bottom: 0` is still required to overcome [editor styles](https://github.com/WordPress/gutenberg/blob/98e96caf8b4668839a4a2a14ade1b37fe272bc1f/packages/interface/src/components/complementary-area/style.scss#L42).
2. The `HStack`'s `div` conflicted with the styles aiming to re-establish a grid layout when slots wrap the fills in a `div`

As I continue to work on switching the panel to use the `Grid` component to handle layout, I'll see if that second issue can't be mitigated.

File: packages/components/src/tools-panel/styles.ts
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/styles.ts]: The dropdown menu classnames have been reinstated and an explanatory comment as to their temporary nature has been added.

**Comment by aaronrobertshaw** [comment]
This has been rebased after the [plus icon changes](https://github.com/WordPress/gutenberg/pull/34107) for the panel menu were merged. 

All the requested changes have been made. Assuming that the [tweak to enforce no bottom margin](https://github.com/WordPress/gutenberg/pull/35415#discussion_r728542197) on the header is acceptable, I think this should be close now.

**Comment by aaronrobertshaw** [comment]
I have a [follow-up PR](https://github.com/WordPress/gutenberg/pull/35621) to this one switching the `ToolsPanel` to leverage the `Grid` component removing the reliance on the `single-column` class. https://github.com/WordPress/gutenberg/pull/35621

File: packages/components/src/tools-panel/styles.ts
Comment (ciampo) [reply packages/components/src/tools-panel/styles.ts]: Thank you for the explanation!

Regarding your second point, I think a better way to write that workardound could be to use the `ToolsPanelHeader` styles object as a selector

<details><summary>Example:</summary>

```diff
diff --git a/packages/components/src/tools-panel/styles.ts b/packages/components/src/tools-panel/styles.ts
index 39fd6e4e74..270dca54e8 100644
--- a/packages/components/src/tools-panel/styles.ts
+++ b/packages/components/src/tools-panel/styles.ts
@@ -38,20 +38,31 @@ export const ToolsPanel = css`
 	padding: ${ space( 4 ) };
 `;
 
+export const ToolsPanelHeading = css`
+	font-size: inherit;
+	font-weight: 500;
+	line-height: normal;
+
+	/* Required to meet specificity requirements to ensure zero margin */
+	&& {
+		margin: 0;
+	}
+`;
+
 /**
  * Items injected into a ToolsPanel via a virtual bubbling slot will require
  * an inner dom element to be injected. The following rule allows for the
  * CSS grid display to be re-established.
  */
 export const ToolsPanelWithInnerWrapper = css`
-	> div:not( :first-of-type ) {
+	> div:not( ${ ToolsPanelHeading } ) {
 		${ toolsPanelGrid.container }
 		${ toolsPanelGrid.item.fullWidth }
 	}
 `;
 
 export const ToolsPanelHiddenInnerWrapper = css`
-	> div:not( :first-of-type ) {
+	> div:not( ${ ToolsPanelHeading } ) {
 		display: none;
 	}
 `;
@@ -81,17 +92,6 @@ export const ToolsPanelHeader = css`
 	}
 `;
 
-export const ToolsPanelHeading = css`
-	font-size: inherit;
-	font-weight: 500;
-	line-height: normal;
-
-	/* Required to meet specificity requirements to ensure zero margin */
-	&& {
-		margin: 0;
-	}
-`;
-
 export const ToolsPanelItem = css`
 	${ toolsPanelGrid.item.fullWidth }
```
</details>

File: packages/components/src/tools-panel/styles.ts
Comment (ciampo) [inline packages/components/src/tools-panel/styles.ts:61]: Bit of a nit, but an alternative approach to this would be to use components composition instead of CSS:

```diff
diff --git a/packages/components/src/tools-panel/styles.ts b/packages/components/src/tools-panel/styles.ts
index 39fd6e4e74..0f63d6f5f8 100644
--- a/packages/components/src/tools-panel/styles.ts
+++ b/packages/components/src/tools-panel/styles.ts
@@ -58,7 +58,6 @@ export const ToolsPanelHiddenInnerWrapper = css`
 
 export const ToolsPanelHeader = css`
 	${ toolsPanelGrid.item.fullWidth }
-	gap: ${ space( 2 ) };
 
 	/**
 	 * The targeting of dropdown menu component classes here is a temporary
diff --git a/packages/components/src/tools-panel/tools-panel-header/component.tsx b/packages/components/src/tools-panel/tools-panel-header/component.tsx
index 4091aed45e..f025a80081 100644
--- a/packages/components/src/tools-panel/tools-panel-header/component.tsx
+++ b/packages/components/src/tools-panel/tools-panel-header/component.tsx
@@ -18,6 +18,7 @@ import MenuGroup from '../../menu-group';
 import MenuItem from '../../menu-item';
 import { HStack } from '../../h-stack';
 import { Heading } from '../../heading';
+import { Spacer } from '../../spacer';
 import { useToolsPanelHeader } from './hook';
 import { contextConnect, WordPressComponentProps } from '../../ui/context';
 import type {
@@ -144,10 +145,11 @@ const ToolsPanelHeader = (
 		: _x( 'View options', 'Button label to reveal tool panel options' );
 
 	return (
-		<HStack { ...headerProps } ref={ forwardedRef }>
+		<HStack { ...headerProps } spacing={ 2 } ref={ forwardedRef }>
 			<Heading level={ 2 } className={ headingClassName }>
 				{ labelText }
 			</Heading>
+			<Spacer />
 			{ hasMenuItems && (
 				<DropdownMenu
 					icon={ dropDownMenuIcon }
```

WDYT? I'm also OK with leaving the code as-is

**Review by ciampo — APPROVED** [review]
Thank you @aaronrobertshaw!

I left a couple more comments, but feel free to merge once they are addressed 🚀 

> I have a follow-up PR to this one switching the ToolsPanel to leverage the Grid component removing the reliance on the single-column class

Excellent, thank you! Just FYI I may not be able to have a look at it today (but I'll try tomorrow) — thank you!

Another nice follow-up to this PR would be to type the `DropdownMenu` component, in order to get rid of the last hardcoded classnames — although I understand if can't look at it now, since other work on ToolsPanel and Typography tools may have higher priority.

File: packages/components/src/tools-panel/styles.ts
Comment (sarayourfriend) [reply packages/components/src/tools-panel/styles.ts]: @ciampo I don't think it's possible to use the serialized styles object as a selector is it? I thought it was just styled components that could be (but I'll be happily surprised if I'm wrong!)

Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/styles.ts]: The suggested change doesn't work for me. No errors are thrown however those styles are omitted.

There might be a way to achieve a valid syntax here that I'm missing. I'm not sure it offers much at this stage though.

File: packages/components/src/tools-panel/styles.ts
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/styles.ts:61]: I'm inclined to leave it as is.

The gap is presentational so sounds like the proper use of CSS to me. I applied the suggested changes to try it out given I've not used `Spacer` previously. It does add additional markup here so I'm missing what the advantage would be.

Given you've also said you're ok with the current code. I'll move forward with it.

**Comment by aaronrobertshaw** [comment]
> Excellent, thank you! Just FYI I may not be able to have a look at it today (but I'll try tomorrow) — thank you!

There's no hurry 🙂 

The switch to using the `Grid` component will make the `ToolsPanel` more flexible in terms of layout so it will be good to work through that one whenever we can.

> Another nice follow-up to this PR would be to type the DropdownMenu component, in order to get rid of the last hardcoded classnames — although I understand if can't look at it now, since other work on ToolsPanel and Typography tools may have higher priority.

It would be a nice follow up. Unfortunately, it will not be something that I can work on. The typography tools have my immediate focus and following that I'll need to return to a greater focus on design tooling for blocks.

Happy to help out with reviews etc if someone else would like to pick up that conversion.

File: packages/components/src/tools-panel/styles.ts
Comment (ciampo) [reply packages/components/src/tools-panel/styles.ts:61]: You make a good point — I also agree with what you said, and using custom CSS here is definitely not an issue.

On the other hand, since I've been working on components for the past ~3 months, I now always try to favour using components "declaratively" (instead of writing custom styles). I guess the vision is that, unless we're creating something particularly custom, we could use existing components as building blocks and avoid adding custom styles.

In this particular case, for example, I was hoping to be able to use only the `spacing` prop without the need to add a `Spacer` (I haven't written those components). Maybe a better solution would have been to use `Flex` instead of `HStack`?

I'm trying to understand how to better use the components that we have to build consistent UIs.

**Comment by ciampo** [comment]
> It would be a nice follow up. Unfortunately, it will not be something that I can work on. The typography tools have my immediate focus and following that I'll need to return to a greater focus on design tooling for blocks.
> Happy to help out with reviews etc if someone else would like to pick up that conversion.

No problems at all! I already really appreciate the help you've been giving with `ToolsPanel`

File: packages/components/src/tools-panel/styles.ts
Comment (ciampo) [reply packages/components/src/tools-panel/styles.ts]: My bad. I tested it by checking it the `WithSlotFillItems` Storybook example thinking that it would cover this case, but I was wrong.

> I thought it was just styled components that could be (but I'll be happily surprised if I'm wrong!)

Me too, but then testing on the wrong Storybook example tricked me into thinking it would work 😅

**Pull request merged**
